arch: arm, arm64: Disable swi_tables.ld file when not required

rakons opened the pull request:
This commit removes the need of swi_tables.ld file if the ISR table generator is not configured to use it.

Fixes #68508

rakons committed:
arch: arm, arm64: Disable swi_tables.ld file when not required

This commit removes the need of swi_tables.ld file if the
ISR table generator is not configured to use it.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

nordicjm commented on arch/arm/core/swi_tables.ld:
```diff
@@ -3,6 +3,6 @@
  */
-#if LINKER_ZEPHYR_FINAL
+#if LINKER_ZEPHYR_FINAL && CONFIG_ISR_TABLES_LOCAL_DECLARATION
 INCLUDE zephyr/isr_tables_swi.ld
```
Fix should be to remove the zephyr prefix since this file will resize in the zephyr folder and would be a double prefix. Ninja (wrongly) adds a workaround for this, make does not, let's make the proper fix

tejlmand replied:
and it still means the LTO compatible ISR feature introduced in #66392 is not working with Makefiles.

We should not have some features only working with a specific generator.

rakons replied:
The zephyr prefix removed.

pdgendt commented on arch/arm/core/swi_tables.ld:
```diff
@@ -3,6 +3,6 @@
  * SPDX-License-Identifier: Apache-2.0
  */
-#if LINKER_ZEPHYR_FINAL
+#if LINKER_ZEPHYR_FINAL && CONFIG_ISR_TABLES_LOCAL_DECLARATION
```
Does this work if `CONFIG_ISR_TABLES_LOCAL_DECLARATION` isn't defined?

```suggestion
#if LINKER_ZEPHYR_FINAL && defined(CONFIG_ISR_TABLES_LOCAL_DECLARATION)
```

rakons replied:
C preprocessor treads undefined symbols as 0 in such situation. defined(...) is here to be able to use undefined preprocessor macros in C conditionals. I can change it anyway.

rakons replied:
Done

rakons committed:
arch: arm, arm64: Remove zephyr prefix from linker includes

This fix removes the zephyr/ prefix from linker included files.
With this prefix the build works only for Ninja and not for
other build tools.

Linking in Zephyr / CMake:
 - Ninja invokes linking directly from <build>.
 - Make invokes linking form <build>/zephyr.

The linker default uses cwd for looking up INCLUDE directives if not found
in list of includes.
Zephyr always adds <build>/zephyr as link include using CMake,
and this is passed to ld as -L<build>/zephyr therefore using
INCLUDE isr_tables_swi.ld ensures it will be correctly found in all cases.

Signed-off-by: Radosław Koppel <radoslaw.koppel@nordicsemi.no>

tejlmand reviewed (changes requested):
For other to understand and for future references, then this is imprecise:
> With this prefix the build works only for Ninja and does not
work propery for other build tools.


Please include something like the following information:
> Linking in Zephyr / CMake.
> - Ninja invokes linking directly from `<build>`.
> - Make invokes linking form `<build>/zephyr`.
>
> The linker default uses cwd for looking up INCLUDE directives if not found in list of includes.
> Zephyr always adds `<build>/zephyr` as link include using CMake,
> and this is passed to ld as `-L<build>/zephyr`  therefore using `INCLUDE isr_tables_swi.ld` ensures it will be correctly found in all cases.

pdgendt commented on arch/arm/core/swi_tables.ld:
```diff
@@ -3,6 +3,6 @@
  */
-#if LINKER_ZEPHYR_FINAL
-INCLUDE zephyr/isr_tables_swi.ld
+#if LINKER_ZEPHYR_FINAL && defined (CONFIG_ISR_TABLES_LOCAL_DECLARATION)
```
Formatting isn't enforced on `.ld` files, but it should try to keep some consistency with the `.clang-format` for c files.

```suggestion
#if LINKER_ZEPHYR_FINAL && defined(CONFIG_ISR_TABLES_LOCAL_DECLARATION)
```

Also for the other changes.

rakons replied:
I also preffer the shorter form. I just aligned my own style here do linker.ld file where it is always with a space after defined.

rakons replied:
Ok... not always - just the nearest the place of my edition

microbuilder commented:
> Please include something like the following information:
> 
> > Linking in Zephyr / CMake.
> > 
> > * Ninja invokes linking directly from `<build>`.
> > * Make invokes linking form `<build>/zephyr`.
> > 
> > The linker default uses cwd for looking up INCLUDE directives if not found in list of includes.
> > Zephyr always adds `<build>/zephyr` as link include using CMake,
> > and this is passed to ld as `-L<build>/zephyr`  therefore using `INCLUDE isr_tables_swi.ld` ensures it will be correctly found in all cases.

I made a comment in the original PR, but these changes are important enough they really need to go into the release notes. A separate PR can be raised for that, though.

rakons commented:
And fixed spaces. But please take a note that we are not consistent about defined and space after it. Personally I prefer it anyway how @pdgendt suggested - just aligned previously my style to the one found in the near surrounding.

rakons commented:
@microbuilder 
> I made a comment in the original PR, but these changes are important enough they really need to go into the release notes. A separate PR can be raised for that, though.

Here is proposal: #68542 68542
